src/net/ossl.cc

```diff
@@ -263,7 +323,21 @@ class tls::certificate_credentials::impl {
     void dh_params(const tls::dh_params&) {}
 
     std::vector<cert_info> get_x509_info() const {
-        return {};
+        std::vector<cert_info> cert_infos;
+        STACK_OF(X509_OBJECT) *chain = X509_STORE_get0_objects(_creds.get());
```

Comment:
@oleiman Does this functionally do the same thing as the GnuTLS analogue in `tls.cc` ?

Reply:
caveat: very limited knowledge of the openssl APIs.

Reply:
That said, I wrote some pretty gnarly integration tests for this stuff, building and deploying a real CA chain and all that. Not sure whether you're able to run those at this stage, but if so it could be good 👍/👎 signal.

Reply:
Maybe I'm wrong but the way I'm reading the gnutls documentation is that the `get_x509_info` function returns the certificates that are paired with keys (that is leaf certificates at the end of the CA chain) while `get_x509_trust_list_info` returns all the certificates in the trusted chain of certificates.

For `get_x509_info` the call to [`gnutls_certificate_get_x509_crt`](https://man7.org/linux/man-pages/man3/gnutls_certificate_get_x509_crt.3.html) returns certificates set via `gnutls_certificate_set_x509_key_mem()` which is how you set the key/cert pair.

For `get_x509_trust_list_info`, the call to [`gnutls_certificate_get_trust_list`](https://man7.org/linux/man-pages/man3/gnutls_certificate_get_trust_list.3.html) returns the list of _trusted_ certificates.  Which I don't believe includes the certificate set via `gnutls_certificate_set_x509_key_mem()`.